src/openrct2/paint/map_element/virtual_floor.c

```diff
@@ -0,0 +1,296 @@
+/*****************************************************************************
+ * OpenRCT2, an open source clone of Roller Coaster Tycoon 2.
+ *
+ * OpenRCT2 is the work of many authors, a full list can be found in contributors.md
+ * For more information, visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * A full copy of the GNU General Public License can be found in licence.txt
+ *****************************************************************************/
+#pragma endregion
+
+#include "map_element.h"
+#include "../paint.h"
+#include "../../input.h"
+#include "../../interface/viewport.h"
+#include "../../sprites.h"
+#include "../../world/map.h"
+
+const rct_xy_element offsets[4] =
+{ 
+    { -32,   0 },
+    {   0,  32 },
+    {  32,   0 },
+    {   0, -32 }
+};
+
+void virtual_floor_paint(paint_session * session)
+{
+        // We only show when the placement modifier keys are active
+    if (!input_test_place_object_modifier(PLACE_OBJECT_MODIFIER_COPY_Z | PLACE_OBJECT_MODIFIER_SHIFT_Z))
+    {
+        return;
+    }
+    
+    rct_xy16 * tile;
+
+    bool    show        = false;
+    bool    weAreLit    = false;
+
+        // Check if map selection (usually single tiles) are enabled
+        //  and if the current tile is near or on them
+    if ((gMapSelectFlags & MAP_SELECT_FLAG_ENABLE) &&
+        session->MapPosition.x >= gMapSelectPositionA.x - gMapVirtualFloorBaseSize &&
+        session->MapPosition.y >= gMapSelectPositionA.y - gMapVirtualFloorBaseSize &&
+        session->MapPosition.x <= gMapSelectPositionB.x + gMapVirtualFloorBaseSize &&
+        session->MapPosition.y <= gMapSelectPositionB.y + gMapVirtualFloorBaseSize)
+    {
+        show = true;
+        if (session->MapPosition.x >= gMapSelectPositionA.x &&
+            session->MapPosition.y >= gMapSelectPositionA.y &&
+            session->MapPosition.x <= gMapSelectPositionB.x &&
+            session->MapPosition.y <= gMapSelectPositionB.y)
+        {
+            weAreLit = true;
+        }
+    }
+
+    if (!show)
+    {
+            // Check if we are anywhere near the selection tiles (larger scenery / rides)
+        if (gMapSelectFlags & MAP_SELECT_FLAG_ENABLE_CONSTRUCT)
+        {
+            for (tile = gMapSelectionTiles; tile->x != -1; tile++)
+            {
+                if (session->MapPosition.x >= tile->x - gMapVirtualFloorBaseSize &&
+                    session->MapPosition.y >= tile->y - gMapVirtualFloorBaseSize &&
+                    session->MapPosition.x <= tile->x + gMapVirtualFloorBaseSize &&
+                    session->MapPosition.y <= tile->y + gMapVirtualFloorBaseSize)
+                {
+                    show = true;
+                }
+            }
+            if (!show)
```

Comment:
you can merge these two tail conditions into one after the `if` block starting in line 63